pilot/pkg/bootstrap/server.go

```diff
@@ -443,7 +449,47 @@ func (c *mockController) Run(<-chan struct{}) {}
 
 // initConfigController creates the config controller in the pilotConfig.
 func (s *Server) initConfigController(args *PilotArgs) error {
-	if args.Config.Controller != nil {
+	mcpServerAddrs := args.MCPServerAddrs
+	if len(mcpServerAddrs) > 0 {
+		// TODO: log random uuid and assign to this pilot node
+		clientNodeID := "some-id"
+		supportedTypes := make([]string, len(model.IstioConfigTypes))
+		for _, model := range model.IstioConfigTypes {
+			supportedTypes = append(supportedTypes, model.MessageName)
+		}
+		log.Infof("mcp client initialized with supported types: %+v\n", supportedTypes)
+		logger := log.RegisterScope("coredatamodel", "mcp controller debugging", 0)
+		mcpController := coredatamodel.NewController(logger)
+		s.addStartFunc(func(stop <-chan struct{}) error {
+			go mcpController.Run(stop)
+			return nil
+		})
+
+		for _, addr := range mcpServerAddrs {
+			conn, err := grpc.Dial(addr, grpc.WithInsecure())
+			if err != nil {
+				log.Infof("Unable connecting to MCP Server: %v\n", err)
+				return err
+			}
+
+			cl := mcp.NewAggregatedMeshConfigServiceClient(conn)
+			mcpClient := mcpclient.New(cl, supportedTypes, mcpController, clientNodeID, map[string]string{})
+
+			s.addStartFunc(func(stop <-chan struct{}) error {
+				ctx, cancel := context.WithCancel(context.Background())
```

Comment:
We can certainly take that approach and we did think about that, however wasn't too sure what would be the desired behaviour, since both approaches have quite different behaviours. Assuming it is ok to share the context among all the mcp clients we can make that change.

Reply:
+ @costinm 

What is the different behavior you were concerned about? My understanding was that derived child contexts would not effect the parent or sibling contexts, e.g. a deadline on client0 wouldn't effect client1.

Reply:
@ayj To further clarify what I meant by the different behaviour, is stop chan received at the same time among all the goroutines [here](https://github.com/istio/istio/blob/d995c62680a7c79a1f0a66cb27d191e2a8ac4e5b/pilot/pkg/bootstrap/server.go#L478)? if so then it makes sense to have one cancel func from a single context to cancel all the routines. However if the stop is going to be received at a different rates then we might want to keep creating a cancel function per routines so this way as single cancel will not cancel everything.

Reply:
The current list of MCP server is fixed at startup. If that list becomes dynamic (via re-loading config) then we'd probably want per-client context.

Reply:
Sure, in that case it makes sense to have a single context to cancel everything, I will update this to reflect the behaviour discussed.